File: 백준/Gold/14891. 톱니바퀴/톱니바퀴.java
import java.util.*;

public class Main {
    static class Gear {
        private int[] teeth = new int[8];
        
        public Gear(String config) {
            for (int i = 0; i < 8; i++) {
                teeth[i] = config.charAt(i) - '0';
            }
        }
        
        public void rotate(int direction) {
            if (direction == 1) { // Clockwise rotation
                int last = teeth[7];
                for (int i = 7; i > 0; i--) {
                    teeth[i] = teeth[i - 1];
                }
                teeth[0] = last;
            } else if (direction == -1) { // Counterclockwise rotation
                int first = teeth[0];
                for (int i = 0; i < 7; i++) {
                    teeth[i] = teeth[i + 1];
                }
                teeth[7] = first;
            }
        }
        
        public int get(int position) {
            return teeth[position];
        }
    }
    
    private static void rotateGears(List<Gear> gears, int wheelNum, int direction) {
        int[] rotate = new int[gears.size()];
        rotate[wheelNum - 1] = direction;
        
        // Check rotations to the left
        for (int i = wheelNum - 1; i > 0; i--) {
            if (gears.get(i).get(6) != gears.get(i - 1).get(2)) {
                rotate[i - 1] = -rotate[i];
            } else {
                break;
            }
        }
        
        // Check rotations to the right
        for (int i = wheelNum - 1; i < gears.size() - 1; i++) {
            if (gears.get(i).get(2) != gears.get(i + 1).get(6)) {
                rotate[i + 1] = -rotate[i];
            } else {
                break;
            }
        }
        
        // Apply rotations
        for (int i = 0; i < gears.size(); i++) {
            gears.get(i).rotate(rotate[i]);
        }
    }
    
    private static int calculateScore(List<Gear> gears) {
        int score = 0;
        int[] points = {1, 2, 4, 8};
        for (int i = 0; i < gears.size(); i++) {
            if (gears.get(i).get(0) == 1) {
                score += points[i];
            }
        }
        return score;
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<Gear> gears = new ArrayList<>();
        
        for (int i = 0; i < 4; i++) {
            String input = scanner.next();
            gears.add(new Gear(input));
        }
        
        int k = scanner.nextInt();
        for (int i = 0; i < k; i++) {
            int wheelNum = scanner.nextInt();
            int direction = scanner.nextInt();
            rotateGears(gears, wheelNum, direction);
        }
        
        System.out.println(calculateScore(gears));
        scanner.close();
    }
}

File: 백준/Gold/14891. 톱니바퀴/톱니바퀴.py
from collections import deque

def rotate_gears(wheels, wheel_num, direction):
    n = len(wheels)
    rotate = [0] * n  # 각 톱니바퀴의 회전 방향을 저장할 배열 (0: 회전 안함, 1: 시계 방향, -1: 반시계 방향)
    rotate[wheel_num - 1] = direction

    # 왼쪽 톱니바퀴들에 대한 회전 전파
    for i in range(wheel_num - 1, 0, -1):
        if wheels[i][6] != wheels[i - 1][2]:
            rotate[i - 1] = -rotate[i]
        else:
            break

    # 오른쪽 톱니바퀴들에 대한 회전 전파
    for i in range(wheel_num - 1, n - 1):
        if wheels[i][2] != wheels[i + 1][6]:
            rotate[i + 1] = -rotate[i]
        else:
            break

    # 모든 계산된 회전 적용
    for i in range(n):
        wheels[i].rotate(rotate[i])

def calculate_score(wheels):
    points = [1, 2, 4, 8]
    score = 0
    for i in range(4):
        if wheels[i][0] == 1:  # S극이면
            score += points[i]
    return score

# 톱니바퀴 상태 입력
wheels = []
for _ in range(4):
    wheel = deque(map(int, input().strip()))
    wheels.append(wheel)

# 회전 횟수와 회전 명령 입력
k = int(input())
for _ in range(k):
    wheel_num, direction = map(int, input().split())
    rotate_gears(wheels, wheel_num, direction)

# 결과 점수 계산 및 출력
print(calculate_score(wheels))